# **(Online Retail Transaction Analysis)**

## Objectives

* Load the Online Retail Transaction dataset into Python using pandas.
* Inspect the raw data and identify data quality issues.
* Clean and transform the dataset by handling missing values, duplicates, invalid records and incorrect data types.
* Creates useful features, including total transaction value.
* Prepare and save a cleaned dataset for exploratory data analysis, machine learning and Power BI visualisation. 

## Inputs

* Raw Online Retail Transaction CSV dataset.
* Online Retail.csv
* Python and pandas libraries.
* Dataset containing transaction information such as invoice number, stock code, Description, quantity, invoice data, unit price, customer ID and country.

## Outputs

* Clean and transformed retail transaction dataset. 
* New calculated feature such as TotalPrice.
* Data quality checks and summry information.
* Cleaned CSV file saved for EDA, machine learning and Power BI.
* Documentation of the data cleaning and transformation process.

## Additional Comments

* This notebook forms the ETL stage of the Online Retail Transaction Analysis project. The raw dataset will be preserved separately from the cleaned dataset. 


---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [2]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\Pdhar\\OneDrive\\Documents\\GitHub\\Online_Retail_Transaction_Analysis\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [3]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [4]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\Pdhar\\OneDrive\\Documents\\GitHub\\Online_Retail_Transaction_Analysis'

# Section 1

Section 1 content

# Extract Data
The raw online Retail Transaction dataset is loaded from the Dataset/RawData folder using pandas.



In [5]:
import os 
import pandas as pd 
import numpy as np

In [6]:
# load the raw dataset
df = pd.read_csv(r"C:\Users\Pdhar\OneDrive\Documents\GitHub\Online_Retail_Transaction_Analysis\Dataset\RawData\Online Retail.csv")

# Display the first five raws
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


### Initial Data Inspection

The raw Dataset has been successfully loaded into pandas DataFrame.
The df.head() output displays the five transactions records and confirms that the dataset has been read successfully.

## Dataset Dimentions

Identify the number of row and columns in the raw dataset.

This shows an overview of the size of dataset before any cleaning or transformation.



In [7]:
# Display the number of rows and columns

print("Number of row:",
df.shape[0])
print("Number of columns:",
      df.shape[1] )

Number of row: 541909
Number of columns: 8


## Columns names
   The columns are examined to understand the variables available in the dataset and determine which fields will be required for the analysis.

In [8]:
print("Column names:")
print(df.columns.tolist())

Column names:
['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


## Columns names findings
These output columns provide the necessary information for analysing retail transactions, sales performance, products, customers and patterns.

## Data types
Correct data types are important for calculation, filtering, date analysis and visualisation.  

In [10]:
df.dtypes

InvoiceNo          str
StockCode          str
Description        str
Quantity         int64
InvoiceDate        str
UnitPrice      float64
CustomerID       int64
Country            str
dtype: object

---

### Data Type Findings

'InvoiceNo', 'StockCode', 'Description', and 'Country, are comes as string values, which is appropriate and identify informatin.
 'Quantity' is stored as 'int64', which is appropriate because it represents the number of items purchased.
 'Unitprice' is store as 'float64'whic is monetary values. 
 'CustomerID' is stored as 'int64'. This will be review during cleaning process.
 'InvoiceDate, is currently store as string. It will be convert to a date time format.


## Missing Values

Missing values are check to identify incomplete records and determine which columns require cleaning before analysis.

In [12]:
df.isnull().sum()

InvoiceNo         0
StockCode         0
Description    1454
Quantity          0
InvoiceDate       0
UnitPrice         0
CustomerID        0
Country           0
dtype: int64

### Missing Values

The result shows that most of the columns contain no missing values. 
However, 'Description' column contains 1454 missing values.



## Duplicate Rows

   Duplicate rows are checked to identify repeated transaction records that could affect the accuracy of the analysis.

In [13]:
df.duplicated().sum()

np.int64(5268)

### Duplicate Rows result
    The dataset contains 5268 duplicate rows. These duplicate records represent repeated transaction and affect the accuracy of the analysis. It will remove during data cleaning.

# Section 2

Section 2 content

---

NOTE

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [9]:
import os
try:
  # create your folder here
  # os.makedirs(name='')
except Exception as e:
  print(e)


IndentationError: expected an indented block after 'try' statement on line 2 (553063055.py, line 5)